File: lib/stack.c
#include <stack.h>
#include <cpu.h>
#include <bus.h>

void stack_push(u8 data)
{
  cpu_get_registers()->sp--;
  bus_write(cpu_get_registers()->sp, data);
}

void stack_push16(u16 data)
{
  stack_push((data >> 8) & 0xFF);
  stack_push(data & 0xFF);
}

u8 stack_pop()
{
  return bus_read(cpu_get_registers()->sp++);
}

u16 stack_pop16()
{
  u16 hi = stack_pop();
  u16 lo = stack_pop();
  return (hi << 8) | lo;
}

File: src/cpu/stack.ts
import { CPU } from "./cpu";

export function stackPush(this: CPU, data: number) {
  this.registers.sp--;
  this.emulator.busWrite(this.registers.sp, data);
}

export function stackPush16(this: CPU, data: number) {
  this.stackPush((data >> 8) & 0xFF);
  this.stackPush(data & 0xFF);
}

export function stackPop(this: CPU) {
  return this.emulator.busRead(this.registers.sp++);
}

export function stackPop16(this: CPU) {
  const low = this.stackPop();
  const high = this.stackPop();

  return (high << 8) | low;
}